### PR TITLE
Slice error tests

### PR DESCRIPTION
**Pull request by hmf**
While documenting the API usage for my notes, I had a problem discussed in issue #11. This PR adds the tests I used when the problem was detected. 

Besides the tests, I noticed a compilation error that was also corrected by adding a `requires` module. See `java/module-info.java` that has this line:

>   requires java.desktop; // required for java.awt.*

- [x] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

File: ndarray/src/main/java/module-info.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
  */
 module org.tensorflow.ndarray {
   requires jdk.unsupported; // required by raw buffer implementations using Unsafe
+  requires java.desktop; // required for java.awt.*
```
Comment (Craigacp) [inline ndarray/src/main/java/module-info.java:19]: This is only used in a test, and I'd prefer to refactor the test to use a serialized byte array than have `java.desktop` as a module level dependency for the whole of TF.

File: ndarray/src/test/java/org/tensorflow/ndarray/IndexTest.java [hunk 2]
```diff
@@ -43,6 +44,371 @@ public void testNullConversions(){
...
+    assertEquals(2, matrix2d.rank());
+
+    /*
+    |(0, 0), (1, 0), (2, 0), (3, 0)|
+    |(0, 1), (1, 1), (2, 1), (3, 1)|
+    |(0, 2), (1, 2), (2, 2), (3, 2)|
+    |(0, 3), (1, 3), (2, 3), (3, 3)|
+    |(0, 4), (1, 4), (2, 4), (3, 4)|  
+    */
+    assertArrayEquals(new String[]{"(0, 0)", "(1, 0)", "(2, 0)", "(3, 0)"}, indexData[0]);
```
Comment (Craigacp) [inline ndarray/src/test/java/org/tensorflow/ndarray/IndexTest.java:67]: This line tests that the test data was generated correctly?

**Review by Craigacp — COMMENTED** [review]
Looks good apart from the module-info update. We'll fix that another way rather than contaminate the runtime module with things that the tests need.

File: ndarray/src/main/java/module-info.java (hunk 1 above)
Comment (hmf) [reply ndarray/src/main/java/module-info.java:19]: Ok, I will remove it.

File: ndarray/src/test/java/org/tensorflow/ndarray/IndexTest.java (hunk 2 above)
Comment (hmf) [reply ndarray/src/test/java/org/tensorflow/ndarray/IndexTest.java:67]: Yes. Just used the first row of the matrix. Excessive? Should I remove it?

Comment (Craigacp) [reply ndarray/src/test/java/org/tensorflow/ndarray/IndexTest.java:67]: No, it's fine. I wanted to check that it was intentional.

File: ndarray/src/main/java/module-info.java (hunk 1 above)
Comment (Craigacp) [reply ndarray/src/main/java/module-info.java:19]: I think we could also supply a module-info in `src/test/java` which has the extra `requires`, that should placate the module system and Maven.

Comment (hmf) [reply ndarray/src/main/java/module-info.java]: @Craigacp Got a minimal `module-info` but needed some adjustments. First, if I stick to the file name `module-info.java`, compilation fails. If I use the file name `module-info.test`, then it compiles and executes the tests correctly. 

The simplest version that I pushed, only adds that one requirement and also changes the module name to `org.tensorflow.ndarray.test`. This last change is not necessary. Keeping the same name, works. As per [this link](https://maven.apache.org/surefire/maven-surefire-plugin/examples/jpms.html), I changed the name.

In the link above, the example seem to have the same entries of the main module repeated in the test module. Should that be done? 

In addition to this, they have (for JUnit5):

```
    requires transitive org.junit.jupiter.engine;
    requires transitive org.junit.jupiter.api;
```

With these, compilation and testing also work. Want to add these too?

Comment (Craigacp) [reply ndarray/src/main/java/module-info.java]: I think using a different module name implies that we won't easily be able to do white box testing (i.e. access protected methods for testing). If it does have a different module name it should have a dependency on the `org.tensorflow.ndarray` module, and either way I think it should have the dependencies on junit.

Comment (hmf) [reply ndarray/src/main/java/module-info.java]: So which is it? Keep the same name or add a dependency?

Comment (Craigacp) [reply ndarray/src/main/java/module-info.java]: I would prefer to use the same name and have Maven patch the module overriding it for the tests. However I've definitely run into issues doing that before, particularly if the test module is depended on elsewhere (which it isn't in this case), so it depends how it goes if you try it. I've not had chance to setup the right environment to test it out locally.

Comment (hmf) [reply ndarray/src/main/java/module-info.java]: I have kept the same module name. Compilation and testing execute correctly. I could not get the `module-info.java` file to work, so its named `module-info.test`. This [reference](https://sormuras.github.io/blog/2018-09-11-testing-in-the-modular-world) did not help but may shed light for anyone trying to change this.

**Review by Craigacp — APPROVED** [review]
LGTM, thanks.

**Pull request merged**